File: BusIk/app/src/main/java/com/example/busik/other/CryptoUtils.java
package com.example.busik.other;

import android.content.Context;
import android.os.Build;
import android.util.Log;

import androidx.annotation.RequiresApi;

import java.io.InputStream;
import java.security.spec.ECField;
import java.util.Base64;
import java.util.Properties;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

public class CryptoUtils {
    private static byte[] key;
    private static byte[] iv;

    @RequiresApi(api = Build.VERSION_CODES.O)
    public static void loadConfig(Context context) {
        try {
            InputStream inputStream = context.getAssets().open("properties.keys");
            Properties props = new Properties();
            props.load(inputStream);

            // Декодирование из Base64
            key = Base64.getDecoder().decode(props.getProperty("aes.key"));
            iv = Base64.getDecoder().decode(props.getProperty("aes.iv"));

            inputStream.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public static String encrypt(String data) {
        try {
            SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
            IvParameterSpec ivSpec = new IvParameterSpec(iv);

            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec);

            byte[] encrypted = cipher.doFinal(data.getBytes());
            return Base64.getEncoder().encodeToString(encrypted);
        }catch (Exception e){
            Log.e("DDD1", "Encrypt error", e);
            return null;
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public static String decrypt(String encryptedData) {
        try {
            SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
            IvParameterSpec ivSpec = new IvParameterSpec(iv);

            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec);

            byte[] decoded = Base64.getDecoder().decode(encryptedData);
            byte[] decrypted = cipher.doFinal(decoded);
            return new String(decrypted);
        }catch (Exception e){
            Log.e("DDD1", "Decrypt error", e);
            return null;
        }
    }
}

File: Operator/OperatorDesktop/src/main/kotlin/com/example/operatordesktop/util/CryptoUtils.kt
package com.example.operatordesktop.util

import java.io.FileInputStream
import java.io.FileOutputStream
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec
import java.util.Base64
import java.util.Properties

object CryptoUtils {
    private const val ALGORITHM = "AES/CBC/PKCS5Padding"
    private var key: ByteArray? = null // 32 символа для AES-256
    private var iv: ByteArray? = null// 16 символов

    // Шифрование
    fun encrypt(phone: String): String {
        val cipher = Cipher.getInstance(ALGORITHM)
        val keySpec = SecretKeySpec(key, "AES")
        val ivSpec = IvParameterSpec(iv)
        cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec)
        val encrypted = cipher.doFinal(phone.toByteArray())
        return Base64.getEncoder().encodeToString(encrypted)
    }

    // Дешифрование
    fun decrypt(encryptedPhone: String): String {
        val cipher = Cipher.getInstance(ALGORITHM)
        val keySpec = SecretKeySpec(key, "AES")
        val ivSpec = IvParameterSpec(iv)
        cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec)
        val decoded = Base64.getDecoder().decode(encryptedPhone)
        val decrypted = cipher.doFinal(decoded)
        return String(decrypted)
    }

    fun generateKeyAndIv(): Pair<ByteArray, ByteArray> {
        val key = ByteArray(32).apply { SecureRandom().nextBytes(this) } // AES-256
        val iv = ByteArray(16).apply { SecureRandom().nextBytes(this) }  // 16 байт для CBC
        return Pair(key, iv)
    }

    fun saveToProperties(key: ByteArray, iv: ByteArray, filePath: String) {
        val props = Properties().apply {
            setProperty("aes.key", Base64.getEncoder().encodeToString(key))
            setProperty("aes.iv", Base64.getEncoder().encodeToString(iv))
        }
        FileOutputStream(filePath).use { output ->
            props.store(output, "AES Configuration")
        }
    }

    fun loadFromProperties(filePath: String): Pair<ByteArray?, ByteArray?> {
        val props = Properties().apply {
            FileInputStream(filePath).use { input -> load(input) }
        }
        key = Base64.getDecoder().decode(props.getProperty("aes.key"))
        iv = Base64.getDecoder().decode(props.getProperty("aes.iv"))
        return Pair(key, iv)
    }

}